File: Arrays/TapeEquilibrium/code.java
// you can also use imports, for example:
import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int[] A) {
        // write your code in Java SE 8
        int N=A.length;
        int Total_sum=0;
        //first find Total sum so that at every index p we can calculate
        // right subarray(Suffix_sum)=Total_sum-(prefix_sum from 0 to p-1 index)
        for(int x:A)
        Total_sum+=x;
        int prefix_Sum=A[0],res=Integer.MAX_VALUE;
        for(int p=1;p<N;p++)
        {
            int suffix_Sum=Total_sum-prefix_Sum;
            res=Math.min(res,Math.abs(prefix_Sum-suffix_Sum));
            prefix_Sum+=A[p];
        }
        return res;

    }
}

File: Caterpiller_Method/MinAbsSumOfTwo/code.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(A) {
    // write your code in JavaScript (Node.js 8.9.4)
    let N = A.length
    let i = 0;
    A.sort((a, b) => a - b);
    while (i < N && A[i] < 0)
        i++;
    let j = i - 1, k = i, sum = Math.abs(A[0] + A[0]);
    while (j >= 0 && i < N) {
        sum = Math.min(sum, Math.abs(A[i] + A[i]));
        sum = Math.min(sum, Math.abs(A[j] + A[i]));
        sum = Math.min(sum, Math.abs(A[j] + A[j]));
        while (j > 0 && A[j] === A[j - 1])
            j--;
        while (i < N - 1 && A[i] === A[i + 1])
            i++;
        if (Math.abs(A[j]) === A[i]) {
            i++;
            j--;

        }
        else if (Math.abs(A[j]) > A[i])
            i++;
        else
            j--;

    }
    while (j >= 0) {
        sum = Math.min(sum, Math.abs(A[j] + A[j]));
        if (j === 0)
            break;
        if (A[j] === A[j - 1]) {
            j--;
        }
        else {
            j--;
        }
    }

    while (i <= N - 1) {
        sum = Math.min(sum, Math.abs(A[i] + A[i]));
        if (i === N - 1) {

            break;
        }
        if (A[i] === A[i + 1]) {

            i++;
        }
        else {

            i++;
        }
    }
    return sum;
}